### PR TITLE
Backend support for retry Solana transactions

### PR DESCRIPTION
**Pull request by yrliou**
- Use current block height + 150 as the last valid block height for transactions with blockhashs not provided by our backend.
- Implement RetryTransaction for Solana, reset transaction data and add a new one. For transactions using durable nonces, re-add the exact transaction for user approval, otherwise, clear the blockhash in the retry transaction and asked for user approval, a new blockhash would be used when user approves.
- Swap transactions cannot be retired as it makes more sense to start over with a new quote from UI flow.
- Partial signed transactions cannot be retried unless using a durable nonce, because dApp needs to re-sign the new transaction too.

Resolves https://github.com/brave/brave-browser/issues/36930

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Try to send swap transaction and dApp transaction, they should be working fine.
If you happen to reproduce transactions not being show up on explorer, check to see if they will be updated to dropped status after about 2 mins or slightly more.

### CODE REVIEW

File: components/resources/wallet_strings.grdp [hunk 1]
```diff
@@ -406,6 +406,9 @@
   <message name="IDS_BRAVE_WALLET_TRANSACTION_PLACEHOLDER" desc="Transactions placeholder">Transactions will appear here</message>
   <message name="IDS_BRAVE_WALLET_TRANSACTION_APPROVE_UNLIMITED" desc="Transaction Approve Unlimited Spend">Unlimited</message>
   <message name="IDS_BRAVE_WALLET_TRANSACTION_NOT_FOUND" desc="Transaction not found error">Transaction not found</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_NOT_RETRIABLE_STATE" desc="Error message indicating that a transaction cannot be retried due to not in a retriable state">This transaction cannot be retried because it's not in rejected, error, or dropped state.</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_NOT_RETRIABLE_TYPE" desc="Error message indicating that a transaction cannot be retried due to its type">This transaction type cannot be retried.</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_PARTIAL_SIGNED" desc="Error message indicating that a transaction is partial signed and cannot be retried from wallet side">This transaction is partially signed and can only be retried from dApp side.</message>
```
Comment (mkarolin) [inline components/resources/wallet_strings.grdp:411]: `partial` -> `partially` in the `desc`

**Review by mkarolin — APPROVED** [review]
`strings`++ with a nit

File: components/resources/wallet_strings.grdp (hunk 1 above)
Comment (yrliou) [reply components/resources/wallet_strings.grdp]: thanks, updated

**Review by nuo-xu — APPROVED** [review]
iOS++ 
created iOS issue to add support retry on Solana tx on iOS https://github.com/brave/brave-browser/issues/37443

File: components/brave_wallet/browser/solana_tx_manager_unittest.cc [hunk 2]
```diff
@@ -429,6 +462,46 @@ class SolanaTxManagerUnitTest : public testing::Test {
...
+        base::BindLambdaForTesting([&](bool success, const std::string& id,
+                                       const std::string& err_message) {
+          EXPECT_EQ(expected_err_message, err_message);
+          EXPECT_EQ(expected_result, success);
+          if (expected_result) {
+            EXPECT_NE(tx_meta_id, id);
+            EXPECT_FALSE(id.empty());
+
+            // Get the new tx and check some fields.
+            auto tx_meta = solana_tx_manager()->GetTxForTesting(id);
```
Comment (darkdh) [inline components/brave_wallet/browser/solana_tx_manager_unittest.cc:489]: can we also check id != tx_meta_id?

**Review by darkdh — APPROVED** [review]

File: components/brave_wallet/browser/solana_tx_manager_unittest.cc (hunk 2 above)
Comment (yrliou) [reply components/brave_wallet/browser/solana_tx_manager_unittest.cc:489]: Yes, there's `EXPECT_NE(tx_meta_id, id);` above.

File: components/brave_wallet/browser/solana_message.cc [hunk 3]
```diff
@@ -691,4 +692,36 @@ std::optional<SolanaMessage> SolanaMessage::FromDeprecatedLegacyValue(
...
+
+  const auto& instruction = instructions_[0];
+  if (instruction.GetAccounts().empty()) {
+    return false;
+  }
+
+  const auto& account = instruction.GetAccounts()[0];
+
+  // Is system program.
+  if (instruction.GetProgramId() != mojom::kSolanaSystemProgramId) {
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/browser/solana_message.cc]: true, removed https://github.com/brave/brave-core/compare/a7083644801898e2db1068bf8383d3ee39d6f532..afbc306c8e4d6f33ad09a93bdff0cf76cda7546c

**Comment by thypon** [comment]
Is the failure final, or can we race a double transaction here? @kdenhartog

**Comment by yrliou** [comment]
@thypon @kdenhartog please see https://bravesoftware.slack.com/archives/C02HSD2FUQJ/p1712850776251649?thread_ts=1710890784.120809&cid=C02HSD2FUQJ for more details.
Here's my understanding:
There are two types of "nonce" used in Solana transactions, 1) blockhash 2) durable nonce.
And the network guarantees the exact transaction will not be accepted more than once. 

1) blockhash: Blockhash is considered expired after 150 blocks, so the previous transaction will no longer be accepted by the network. In this case, we prompt and ask user to re-approve, if approves we will use a new blockhash and sign the new retry transaction. Shouldn't have double spent issue cuz the previous transaction will not be accepted due to blockhash expiration.
2) durable nonce: blockhash field will be the current nonce from the nonce account when you created the transaction. The first instruction in the transaction will always be "AdvanceNonceAccount", so if your transaction is ever accepted by the network already, the exact transaction will not be accepted again cuz the nonce is already advanced and the one in the existing transaction is no longer valid. In this case we prompt and ask user to re-approve, and just sign the exact transaction (hence signature and everything is still the same) and resend. No double spent due to "AdvanceNonceAccount".

**Comment by kdenhartog** [comment]
> Is the failure final, or can we race a double transaction here?

This would only happen if we used blockhash first and retried with the durable nonce method before the blockhash expired. I don't believe that's what is happening here, but I can double check.

File: components/brave_wallet/browser/solana_tx_manager.cc [hunk 4]
```diff
@@ -334,7 +366,69 @@ void SolanaTxManager::SpeedupOrCancelTransaction(
 
 void SolanaTxManager::RetryTransaction(const std::string& tx_meta_id,
```
Comment (kdenhartog) [inline components/brave_wallet/browser/solana_tx_manager.cc:367]: Retry logic LGTM

**Review by kdenhartog — APPROVED** [review]
confirmed that this isn't switching nonce types, so the network would detect any odd state transitions for us as a final check to the logic implemented here.

**Comment by kdenhartog** [comment]
Thanks for also catching the edge cases around swaps and multi sigs. I hadn't thought about the impact of that until I read the code.

**Pull request merged**